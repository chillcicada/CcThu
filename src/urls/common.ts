// #region urls
export const LEARN_PREFIX = 'https://learn.tsinghua.edu.cn'
export const ID_PREFIX = 'https://id.tsinghua.edu.cn'
export const REGISTRAR_PREFIX = 'https://zhjw.cic.tsinghua.edu.cn'
export const CARD_PREFIX = 'https://card.tsinghua.edu.cn'
export const TOJ_PREFIX = 'https://dsa.cs.tsinghua.edu.cn/oj'
export const INFO2021_PREFIX = 'https://info2021.tsinghua.edu.cn'
export const NETWORK_PREFIX = 'https://usereg.tsinghua.edu.cn'
export const WEBVPN_PREFIX = 'https://webvpn.tsinghua.edu.cn'
export const HOME_PREFIX = 'http://myhome.tsinghua.edu.cn' // ! not https
export const OVERLEAF_PREFIX = 'https://overleaf.tsinghua.edu.cn'
/**
 * @deprecated Use INFO2021_PREFIX instead
 */
export const ACADEMIC_PREFIX = 'https://academic.tsinghua.edu.cn'
/**
 * The prefix of the mail system
 * @param isStudent Whether the user is a student, default to `true`
 */
export const MAIL_PREFIX = (isStudent = true) => `https://mail${isStudent ? 's' : ''}.tsinghua.edu.cn`
// #endregion

// #region webvpn urls
// This only works for those that require web vpn when accessing from outside Tsinghua network
// ! This is all under `https` proxy
export const WEBVPN_ID_PREFIX = `${WEBVPN_PREFIX}/https/77726476706e69737468656265737421f9f30f8834396657761d88e29d51367bcfe7`
export const WEBVPN_EMAIL_PREFIX = `${WEBVPN_PREFIX}/https/77726476706e69737468656265737421fdf64890347e7c4377068ea48d546d3011ff591d40`
export const WEBVPN_LIBRARY_ROOM_BOOK_PREFIX = `${WEBVPN_PREFIX}/https/77726476706e69737468656265737421f3f643d22b396a1e6a1b80a29f5d363409e413829737d1'`
export const WEBVPN_NETWORK_PREFIX = `${WEBVPN_PREFIX}/https/77726476706e69737468656265737421e5e4448e223726446d0187ab9040227b54b6c80fcd73`
// #endregion
