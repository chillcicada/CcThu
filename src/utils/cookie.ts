// TODO: optimize the code
import { readFile, unlink, writeFile } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

// ? or use the `.temp` path at workspace root
const TmpPath = tmpdir()
const TargetFileName = 'cookie.tmp'

interface commonCookieConfig {
  path: string
  fileName: string
}

/**
 * Configuration for set cookie
 */
export interface setCookieConfig extends commonCookieConfig {
  cookie: string
}

/**
 * Async set cookie to file
 * @param cfg The configuration to set the cookie,
 * type {@link setCookieConfig | string}, default `{}`,
 * if the configuration is a string, then it will be treated as the cookie
 * @param cfg.cookie The cookie to set, type `string`
 * @param cfg.path The path to save the cookie, default {@link TmpPath}
 * @param cfg.fileName The file name to save the cookie, default {@link TargetFileName}
 *
 * @returns Promise
 */
export async function setCookie(cfg: Partial<setCookieConfig> | string = {}): Promise<void> {
  if (typeof cfg === 'string')
    cfg = { cookie: cfg }

  const {
    cookie = '',
    path = TmpPath,
    fileName = TargetFileName,
  } = cfg

  if (!cookie)
    console.warn('Empty cookie')

  const cookieFile = join(path, fileName)

  return new Promise<void>((resolve, reject) => {
    writeFile(cookieFile, cookie, e => e ? reject(e) : resolve())
  })
}

/**
 * Configuration for get cookie
 */
export interface getCookieConfig extends commonCookieConfig {
  encoding: BufferEncoding
}

/**
 * Async read cookie from file
 * @param cfg The configuration to read the cookie,
 * type {@link getCookieConfig | BufferEncoding}, default `{}`,
 * if the configuration is a {@link BufferEncoding}, then it will be treated as the encoding
 * @param cfg.path The path to read the cookie, default {@link TmpPath}
 * @param cfg.encoding The encoding to read the file, type {@link BufferEncoding}, default `utf-8`
 * @param cfg.fileName The file name to read the cookie, default {@link TargetFileName}
 *
 * @returns The cookie as string
 */
export async function getCookie(cfg: Partial<getCookieConfig> | BufferEncoding = {}): Promise<string> {
  if (typeof cfg === 'string')
    cfg = { encoding: cfg }

  const {
    path = TmpPath,
    encoding = 'utf-8',
    fileName = TargetFileName,
  } = cfg

  const cookieFile = join(path, fileName)

  return new Promise<string>((resolve, reject) => {
    readFile(cookieFile, encoding, (e, cookie) => e ? reject(e) : resolve(cookie))
  })
}

/**
 * Configuration for remove cookie
 */
export interface rmCookieConfig extends commonCookieConfig {}

/**
 * Async write cookie to file
 * @param cfg The configuration to remove the cookie,
 * type {@link rmCookieConfig}, default `{}`
 * @param cfg.path The path to remove the cookie, default {@link TmpPath}
 * @param cfg.fileName The file name to remove the cookie, default {@link TargetFileName}
 */
export async function rmCookie(cfg: Partial<rmCookieConfig> = {}): Promise<void> {
  const {
    path = TmpPath,
    fileName = TargetFileName,
  } = cfg

  const cookieFile = join(path, fileName)

  return new Promise<void>((resolve, reject) => {
    unlink(cookieFile, e => e ? reject(e) : resolve())
  })
}
