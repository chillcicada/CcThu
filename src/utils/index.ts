export {
  $,
  addCSRFTokenToUrl,
  decodeHTML,
  parseSemesterType,
  fetchWithRetry,
  fetchWithTimeout,
  paramToModule,
} from './common'

export { getCookie, rmCookie, setCookie } from './cookie'
