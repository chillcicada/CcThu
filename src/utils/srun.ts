function preprocess(str: string, useLen = false): number[] {
  const v: number[] = []
  const len = str.length

  for (let i = 0; i < len; i += 4) {
    v.push(
      str.charCodeAt(i) | (str.charCodeAt(i + 1) << 8) | (str.charCodeAt(i + 2) << 16) | (str.charCodeAt(i + 3) << 24),
    )
  }

  return useLen ? [...v, len] : v
}

/**
 * xxtea encrypt for srun network
 */
export function xEncode(str: string, key: string): string {
  if (str === '')
    return ''

  const s = preprocess(str, true)
  const k = preprocess(key)

  if (k.length < 4)
    k.length = 4

  const n = s.length - 1
  const delta = 0x9E3779B9

  let z = s[n]
  let y = s[0]
  let q = Math.floor(6 + 52 / (n + 1))
  let d = 0

  while (q-- > 0) {
    d += delta
    const e = d >>> 2 & 3

    for (let p = 0; p < n; p++) {
      y = s[p + 1]
      let mx = z >>> 5 ^ y << 2
      mx += y >>> 3 ^ z << 4 ^ (d ^ y)
      mx += k[p & 3 ^ e] ^ z
      s[p] += mx
      z = s[p]
    }

    y = s[0]
    let mx = z >>> 5 ^ y << 2
    mx += y >>> 3 ^ z << 4 ^ (d ^ y)
    mx += k[n & 3 ^ e] ^ z
    s[n] += mx
    z = s[n]
  }

  return s.map(v => String.fromCharCode(v & 0xFF, v >> 8 & 0xFF, v >> 16 & 0xFF, v >> 24 & 0xFF)).join('')
}

/**
 * A minimal modified base64 encoder/decoder
 */

export function xEncodeBase64(str: string) {
  const _base64Alphabet = 'LVoJPiCN2R8G90yg+hmFHuacZ1OWMnrsSTXkYpUq/3dlbfKwv6xztjI7DeBE45QA'
  const base64Alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'

  return btoa(str).split('').map(c => c === '=' ? '=' : _base64Alphabet.charAt(base64Alphabet.indexOf(c))).join('')
}
