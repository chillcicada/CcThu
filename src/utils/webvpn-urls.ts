import { WEBVPN_ID_PREFIX, WEBVPN_LIBRARY_ROOM_BOOK_PREFIX, WEBVPN_PREFIX } from '@/constants'

/**
 * webvpn login url
 */
export const WebvpnLogin = `${WEBVPN_PREFIX}/login`

/**
 * webvpn oauth login url
 */
export const WebvpnOauthLogin = `${WEBVPN_PREFIX}/oauth/login`

/**
 * webvpn logout url
 */
export const WebvpnLogout = `${WEBVPN_PREFIX}/logout`

/**
 * webvpn id form url
 */
export const WebvpnIdForm = `${WEBVPN_ID_PREFIX}/do/off/ui/auth/login/form/`

/**
 * webvpn id login url
 */
export const WebvpnIdLogin = `${WEBVPN_ID_PREFIX}/do/off/ui/auth/login/check`

// #region library room book
/**
 * webvpn library room book search url
 */
// export const LIBRARY_FUZZY_SEARCH_ID_URL = `${WEBVPN_LIBRARY_ROOM_BOOK_PREFIX}/ic-web/account/getMembers?page=1&pageNum=10&key=`

/**
 * webvpn library room book action url
 */
// export const WebvpnLibraryRoomBookAction = `${WEBVPN_LIBRARY_ROOM_BOOK_PREFIX}/ic-web/reserve`

/**
 * webvpn library room book record url
 * ? what the hell of the query params ??
 */
export const WebvpnLibraryRoomBookRecord = `${WEBVPN_LIBRARY_ROOM_BOOK_PREFIX}/ic-web/reserve/resvInfo?needStatus=8454&orderKey=gmt_create&orderModel=desc`

/**
 * webvpn library room cancel book url
 */
export const WebvpnLibraryRoomCancelBook = `${WEBVPN_LIBRARY_ROOM_BOOK_PREFIX}/ic-web/reserve/delete`

/**
 * webvpn library room update email url
 */
export const WebvpnLibraryRoomUpdateEmail = `${WEBVPN_LIBRARY_ROOM_BOOK_PREFIX}/ic-web/account/update`
// #endregion
