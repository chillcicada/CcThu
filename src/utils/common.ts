import { decodeHTML as _decodeHTML } from 'entities'
import * as cheerio from 'cheerio'
import type { Element } from 'domhandler'
// import type { Element } from 'cheerio'
// // is it possible to import Element from cheerio?ðŸ˜Ž

import { getCookie } from './cookie'

import type { BaseResponse, SemesterType } from '@/types'
import { DEFAULT_RETRY_INTERVAL, DEFAULT_TIMEOUT, MAX_RETRY } from '@/constants'

/**
 * Add a CSRF token to a URL
 * @param url - The URL to add the CSRF token to
 * @param token - The CSRF token to add
 *
 * @returns The {@link URL} with the CSRF token added
 */
export function addCSRFTokenToUrl(url: string | URL, token: string): string {
  if (typeof url === 'string')
    url = new URL(url)
  url.searchParams.set('_csrf', token)
  return url.toString()
}

/**
 * Decode HTML string and Remove the overflow character of the conversion
 * @param html - The HTML string to decode
 *
 * @returns The decoded string
 */
export const decodeHTML = (html: string) => _decodeHTML(html).replace(/^(\\xC2\\x9E\\xC3\\xA9|(\\x9E)?\\xE9)\\x65/, '')

export const noLogin = (res: Response) => res.url.includes('login_timeout') || res.status === 403

/**
 * Turn a param string to a module name
 */
export const paramToModule = (params: string) => params.replace(/\//g, '-')

/**
 * Parse SemesterType from a number
 * @param n - The number to parse
 *
 * @returns The {@link SemesterType} parsed
 */
export function parseSemesterType(n: number): SemesterType {
  switch (n) {
    case 1:
      return 'autumn'
    case 2:
      return 'spring'
    case 3:
      return 'summer'
    default:
      return 'unknown'
  }
}

/**
 * Parse HTML string to Cheerio object
 * @param html - The HTML string or {@link Element} or Element[] to parse
 *
 * @returns The Cheerio object {@link cheerio.CheerioAPI}
 */
export const $ = (html: string | Element | Element[]) => cheerio.load(html, { xml: true, decodeEntities: false })

// #region fetch
/**
 * Config for {@link fetchWithTimeout}
 */
export interface FetchWithTimeoutConfig {
  useCookie: boolean
  useCSRFToken: boolean
  timeout: number
  fetchImpl: typeof fetch
}

/**
 * A wrapper for {@link fetch} with timeout
 * @param url - The string or {@link URL} or {@link Request} to fetch
 * @param options - The fetch options, default to {}
 * @param config - The fetch configuration, type {@link FetchWithTimeoutConfig}
 * @param config.useCookie - Whether to use cookie, default to true
 * @param config.useCSRFToken - Whether to use CSRF token, default to true
 * @param config.timeout - The timeout of the fetch, default to {@link DEFAULT_TIMEOUT}
 * @param config.fetchImpl - The fetch implementation, default to {@link fetch}
 *
 * @returns The {@link Response} of the fetch
 */
export async function fetchWithTimeout(
  url: string | URL | Request,
  options: RequestInit = {},
  config: Partial<FetchWithTimeoutConfig> = {},
): Promise<Response> {
  const {
    useCookie = true,
    timeout = DEFAULT_TIMEOUT,
    fetchImpl = fetch,
  } = config

  // TODO: Optimize this
  if (useCookie) {
    const useCSRFToken = config.useCSRFToken ?? true

    const cookie = await getCookie()

    const defaultOptions = { headers: { cookie } } as const

    options = {
      ...defaultOptions,
      ...options,
    }

    const csrfToken = cookie.match(/XSRF-TOKEN=([^;]+)/)?.[1] || ''

    if (url instanceof Request)
      url = url.url

    if (useCSRFToken)
      url = addCSRFTokenToUrl(url, csrfToken)
  }

  if (timeout === 0)
    return fetchImpl(url, options)

  return new Promise((resolve, reject) => {
    const controller = new AbortController()
    const id = setTimeout(() => {
      controller.abort()
      reject(new Error('Timeout'))
    }, timeout)

    fetchImpl(url, { ...options, signal: controller.signal })
      .then(resolve)
      .catch(reject)
      .finally(() => clearTimeout(id))
  })
}

/**
 * Config for {@link fetchWithRetry}
 */
export interface FetchWithRetryConfig extends FetchWithTimeoutConfig {
  maxRetry: number
  retryInterval: number
}

/**
 * A wrapper for {@link fetch} with timeout and retry
 * @param url - The string or {@link URL} or {@link Request} to fetch
 * @param options - The fetch options, default to {}
 * @param config - The fetch configuration, type {@link FetchWithRetryConfig}
 * @param config.maxRetry - The maximum number of retries, default to {@link MAX_RETRY}
 * @param config.retryInterval - The interval between retries, default to {@link DEFAULT_RETRY_INTERVAL}
 * @param config.timeout - The timeout of the fetch, default to {@link DEFAULT_TIMEOUT}
 * @param config.fetchImpl - The fetch implementation, default to {@link fetch}
 *
 * @returns The {@link Response} of the fetch
 */
export async function fetchWithRetry(
  url: string | URL | Request,
  options: RequestInit = {},
  config: Partial<FetchWithRetryConfig> = {},
): Promise<Response> {
  const {
    maxRetry = MAX_RETRY,
    retryInterval = DEFAULT_RETRY_INTERVAL,
    ...rest
  } = config

  return new Promise((resolve, reject) => {
    let retry = 0

    const fetcher = async () => {
      try {
        resolve(await fetchWithTimeout(url, options, rest))
      }
      catch (e) {
        retry < maxRetry
          ? setTimeout(fetcher, retryInterval) && retry++
          : reject(e)
      }
    }

    fetcher()
  })
}
// #endregion

// TODO
export function SetResponse<T>(res: Response): Promise<T> {
  return res.json()
}

/**
 * Create a success message for {@link BaseResponse}
 * @param message - The message
 */
export const useFail = (message: string): BaseResponse => ({ status: false, message })

/**
 * Create a fail message for {@link BaseResponse}
 * @param e - The error to handle
 * @param message - The message
 * @param callback - The callback to run after handling the error
 */
export function useError(e: unknown, message: string, callback: (e: unknown) => void = () => { }): BaseResponse {
  if (import.meta.env.NODE_ENV === 'development')
    console.error(e)

  callback(e)

  return useFail(message)
}

/**
 * A simple network test
 */
export function netTest(): BaseResponse<string> {
  return {
    message: 'Test Successfully',
    data: new Date().toISOString(),
    status: true,
  }
}
