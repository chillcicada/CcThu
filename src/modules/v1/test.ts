import { useTest } from '@/utils'

export default useTest
