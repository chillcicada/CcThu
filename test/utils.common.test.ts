import { describe, expect, it } from 'bun:test'

import * as c from '@/utils/common'

describe('utils/common', () => {
  it('should add CSRF token to URL', () => {
    const token = 'csrf-token'

    const url1 = 'https://example.com'
    const result1 = 'https://example.com/?_csrf=csrf-token'

    expect(c.addCSRFTokenToUrl(url1, token).toString()).toBe(result1)

    const url2 = 'https://example.com?foo=bar'
    const result2 = 'https://example.com/?foo=bar&_csrf=csrf-token'

    expect(c.addCSRFTokenToUrl(url2, token).toString()).toBe(result2)

    const url3 = 'https://example.com?foo=bar&_csrf=token'
    const result3 = 'https://example.com/?foo=bar&_csrf=csrf-token'

    expect(c.addCSRFTokenToUrl(url3, token).toString()).toBe(result3)
  })

  it('should get CSRF token from URL', () => {
    const targetToken = 'csrf-token'

    const url1 = `https://example.com/?_csrf=${targetToken}`
    expect(c.getCSRFTokenFromUrl(url1)).toBe(targetToken)

    const url2 = `https://example.com/?foo=bar&_csrf=${targetToken}`
    expect(c.getCSRFTokenFromUrl(url2)).toBe(targetToken)

    const url3 = 'https://example.com/?foo=bar'
    expect(c.getCSRFTokenFromUrl(url3)).toBe('')

    const url4 = 'https://example.com'
    expect(c.getCSRFTokenFromUrl(url4)).toBe('')
  })

  it('should parse semester type', () => {
    expect(c.parseSemesterType(1)).toBe('autumn')
    expect(c.parseSemesterType(2)).toBe('spring')
    expect(c.parseSemesterType(3)).toBe('summer')
    expect(c.parseSemesterType(4)).toBe('unknown')
  })
})
