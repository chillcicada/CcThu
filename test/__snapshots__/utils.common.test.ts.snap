// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils/common > should use error 1`] = `
{
  "message": "string error",
  "status": false,
}
`;

exports[`utils/common > should use error 2`] = `
{
  "message": "error",
  "status": false,
}
`;

exports[`utils/common > should use error 3`] = `
{
  "message": "An error occurred!",
  "status": false,
}
`;

exports[`utils/common > should use error 4`] = `
{
  "message": "error msg",
  "status": false,
}
`;

exports[`utils/common > should use fail 1`] = `
{
  "message": "string error",
  "status": false,
}
`;
